### PR TITLE
Introduce a Webpack plugin for re-exporting the lib folder from a bundle.

### PR DESCRIPTION
**Pull request by iclanton**
## Summary

Introduces a plugin specifically for accomplishing what's in this PR: https://github.com/microsoft/rushstack/pull/3840

## How it was tested

Tested with the change on https://github.com/microsoft/rushstack/pull/3840

### CODE REVIEW

File: webpack/webpack-deep-imports-compat-plugin/src/DeepImportsCompatPlugin.ts [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+        for (const libPath of libPaths) {
+          const depth: number = countSlashes(libPath);
+          const requirePath: string = '../'.repeat(depth) + libOutFolderRelativeOutputPath;
+          const moduleText: string = [
+            `module.exports = require('${requirePath}').getPath('${libPath}');`
+          ].join('\n');
+
+          compilation.emitAsset(
+            `${outputPathRelativeLibOutFolder}/${libPath}`,
+            new compiler.webpack.sources.RawSource(moduleText)
```
Comment (TheLarkInn) [inline webpack/webpack-deep-imports-compat-plugin/src/DeepImportsCompatPlugin.ts:265]: If this is really expensive, would it be prudent to use [CachedSource](https://github.com/webpack/webpack-sources#cachedsource)?

**Review by TheLarkInn — CHANGES_REQUESTED** [review]
Have a few comments. Overall this looks mostly good. I'm not 100% sure the exact scenario we are trying to solve with this plugin so if that can be explained in at least one central place that would be good.

**Comment by TheLarkInn** [comment]
After seeing the tied PR (which I didn't notice at first, I think its a bit more clear what the purpose of this plugin is) It would be good to document a small example, etc. in the README so more people understand what we are leveraging it.

File: webpack/webpack-deep-imports-compat-plugin/src/DeepImportsCompatPlugin.ts [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+const HAS_BEEN_APPLIED_SYMBOL: unique symbol = Symbol(`${PLUGIN_NAME}.hasBeenApplied`);
+const JS_EXTENSION: string = '.js';
+
+interface IExtendedConfiguration extends Configuration {
+  [HAS_BEEN_APPLIED_SYMBOL]?: true;
+}
+
+function countSlashes(str: string): number {
```
[the remark this answers is not in the document]
Comment (iclanton) [reply webpack/webpack-deep-imports-compat-plugin/src/DeepImportsCompatPlugin.ts:34]:
```suggestion
/**
 * Returns the number of `/` characters present in a given string.
 */
function countSlashes(str: string): number {
```

File: webpack/webpack-deep-imports-compat-plugin/src/DeepImportsCompatPlugin.ts [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+    webpackConfiguration.plugins.push(virtualModulesPlugin);
+    webpackConfiguration.plugins.push(plugin);
+
+    if (!webpackConfiguration.entry) {
+      webpackConfiguration.entry = {};
+    }
+
+    if (typeof webpackConfiguration.entry === 'string') {
+      throw new Error(`The "entry" option must not be a string.`);
```
[the remark this answers is not in the document]
Comment (iclanton) [reply webpack/webpack-deep-imports-compat-plugin/src/DeepImportsCompatPlugin.ts:135]: Do webpack errors make sense in this the context of the `applyToWebpackConfiguration` function?

File: webpack/webpack-deep-imports-compat-plugin/src/DeepImportsCompatPlugin.ts [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+          }
+        }
+      } else {
+        libPaths = includePaths;
+      }
+
+      libPaths = libPaths.sort();
+
+      const lines: string = [
+        'export function getPath(p) {',
```
[the remark this answers is not in the document]
Comment (iclanton) [reply webpack/webpack-deep-imports-compat-plugin/src/DeepImportsCompatPlugin.ts:191]: It feels like the spread operator would be more awkward in a multiline string.

File: webpack/webpack-deep-imports-compat-plugin/src/DeepImportsCompatPlugin.ts (hunk 1 above)
Comment (iclanton) [reply webpack/webpack-deep-imports-compat-plugin/src/DeepImportsCompatPlugin.ts:265]: That actually feels like it'd probably add more overhead, honestly.

File: webpack/webpack-deep-imports-compat-plugin/README.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    excludePatterns: ['**/*.test.*', '**/test/**/*', '**/__mocks__/**/*']
+  },
+  // The name of the folder to which the loose files should be output to
+  outFolderName: 'lib'
+});
+
+module.exports = configuration
+```
+
+**DO NOT MANUALLY ADD THIS PLUGIN TO YOUR WEBPACK CONFIGURATION'S `plugins` PROPERTY**
```
Comment (iclanton) [inline webpack/webpack-deep-imports-compat-plugin/README.md:116]:
```suggestion
Manually adding this plugin to your Webpack configuration's `plugins` array will result in an error.
```

File: webpack/webpack5-module-minifier-plugin/LICENSE [hunk 6]
```diff
@@ -1,4 +1,4 @@
-@rushstack/webpack5-module-minifier-plugin@5
+@rushstack/webpack5-module-minifier-plugi
```
Comment (iclanton) [inline webpack/webpack5-module-minifier-plugin/LICENSE:1]:
```suggestion
@rushstack/webpack5-module-minifier-plugin
```

**Review by TheLarkInn — APPROVED** [review]
LGTM

**Pull request merged**